cirq-core/cirq/experiments/readout_confusion_matrix.py

```diff
@@ -0,0 +1,310 @@
+            func, result, method='SLSQP', constraints=constraints, bounds=bounds
+        )
+        return res.x
+
+    def __repr__(self) -> str:
+        return (
+            f"cirq.ReadoutConfusionMatrix("
+            f"[{','.join([proper_repr(cm) for cm in self.confusion_matrices])}],"
+            f"{self.measure_qubits}"
+            f")"
+        )
+
+    def _json_dict_(self) -> Dict[str, Any]:
+        return {
+            'confusion_matrices': self.confusion_matrices,
+            'measure_qubits': self.measure_qubits,
+        }
+
+    @classmethod
+    def _from_json_dict_(
+        cls, confusion_matrices, measure_qubits, **kwargs
+    ) -> 'ReadoutConfusionMatrix':
+        return cls([np.asarray(cm) for cm in confusion_matrices], measure_qubits)
+
+    def _approx_eq_(self, other: Any, atol: float) -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self.qubits == other.qubits and all(
+            np.allclose(cm, ocm, atol=atol)
+            for cm, ocm in zip(self.confusion_matrices, other.confusion_matrices)
+        )
+
+    def __eq__(self, other: Any) -> bool:
+        if not isinstance(other, type(self)):
+            return NotImplemented
+        return self.qubits == other.qubits and all(
+            np.array_equal(cm, ocm)
+            for cm, ocm in zip(self.confusion_matrices, other.confusion_matrices)
+        )
+
+    def __ne__(self, other: Any) -> bool:
+        return not self == other
+
+    def __hash__(self) -> int:
+        vals = tuple(v for cm in self.confusion_matrices for _, v in np.ndenumerate(cm))
+        return hash((ReadoutConfusionMatrix, vals, tuple(self.qubits)))
+
+
+def measure_confusion_matrix(
+    sampler: 'cirq.Sampler',
+    qubits: Union[Sequence['cirq.Qid'], Sequence[Sequence['cirq.Qid']]],
+    repetitions: int = 1000,
+) -> ReadoutConfusionMatrix:
+    """Prepares `ReadoutConfusionMatrix` for the n qubits in the input.
+
+    The confusion matrix (CM) for two qubits is the following matrix:
+
+        ⎡ Pr(00o|00a) Pr(01o|00a) Pr(10o|00a) Pr(11o|00a) ⎤
+        ⎢ Pr(00o|01a) Pr(01o|01a) Pr(10o|01a) Pr(11o|01a) ⎥
+        ⎢ Pr(00o|10a) Pr(01o|10a) Pr(10o|10a) Pr(11o|10a) ⎥
+        ⎣ Pr(00o|11a) Pr(01o|11a) Pr(10o|11a) Pr(11o|11a) ⎦
+
+    where Pr(ij | pq) = Probability of observing “ij” given state “pq” was prepared.
+
+    Args:
+        sampler: Sampler to collect the data from.
+        qubits: Qubits for which the confusion matrix should be measured.
+        repetitions: Number of times to sample each circuit for a confusion matrix row.
+    """
+    qubits = cast(
+        Sequence[Sequence['cirq.Qid']], [qubits] if isinstance(qubits[0], ops.Qid) else qubits
+    )
+    confusion_matrices = []
+    for qs in qubits:
+        results = sampler.run_batch(
+            [
+                circuits.Circuit(
+                    [ops.X(q) ** ((state >> i) & 1) for i, q in enumerate(qs[::-1])],
```

Comment:
don't be fancy like this; it's basically unreadable

Reply:
Modified.